modules/dnn/test/test_onnx_confromance.cpp

```diff
@@ -0,0 +1,1060 @@
+
+    std::stringstream ss;
+    ss << test_case.name << "_";
+    PrintTo(backend, &ss);
+    ss << "_";
+    PrintTo(target, &ss);
+
+    return ss.str();
+}
+
+template<typename TString>
+static std::string _tf(TString filename, bool required = true)
+{
+    return findDataFile(std::string("dnn/onnx/") + filename, required);
+}
+
+std::vector<TestCaseInput> readTestCases()
+{
+    std::vector<TestCaseInput> ret;
+    for (const auto& test_case : testConformanceConfig)
+    {
+        TestCaseInput input;
+
+        std::string prefix = cv::format("conformance/node/%s", test_case.name);
+        input.name = test_case.name;
+        input.model_path = _tf(cv::format("%s/model.onnx", prefix.c_str()));
+
+        for (int i = 0; i < test_case.inputs; ++i)
+        {
+            input.input_paths.push_back(_tf(cv::format("%s/test_data_set_0/input_%d.pb", prefix.c_str(), i)));
+        }
+
+        for (int i = 0; i < test_case.outputs; ++i)
+        {
+            input.output_paths.push_back(_tf(cv::format("%s/test_data_set_0/output_%d.pb", prefix.c_str(), i)));
+        }
+
+        ret.emplace_back(std::move(input));
+    }
+
+    return ret;
+}
+
+class Test_ONNX_conformance : public TestWithParam<ONNXConfParams>
+{
+public:
+    TestCaseInput test_case;
+    Backend backend;
+    Target target;
+
+    double default_l1;
+    double default_lInf;
+
+    Test_ONNX_conformance()
+    {
+        auto net_params = std::tie(backend, target);
+        std::tie(test_case, net_params) = GetParam();
+
+        if (target == DNN_TARGET_CUDA_FP16 || target == DNN_TARGET_OPENCL_FP16 || target == DNN_TARGET_MYRIAD)
+        {
+            default_l1 = 4e-3;
+            default_lInf = 2e-2;
+        }
+        else
+        {
+            default_l1 = 1e-5;
+            default_lInf = 1e-4;
+        }
+    }
+
+    void expectNoFallbacks(Net& net) const
+    {
+        // Check if all the layers are supported with current backend and target.
+        // Some layers might be fused so their timings equal to zero.
+        std::vector<double> timings;
+        net.getPerfProfile(timings);
+        std::vector<String> names = net.getLayerNames();
+        ASSERT_EQ(names.size(), timings.size());
+
+        bool hasFallbacks = false;
+        for (int i = 0; i < names.size(); ++i)
+        {
+            Ptr<dnn::Layer> l = net.getLayer(net.getLayerId(names[i]));
+            bool fused = timings[i] == 0.;
+            if ((!l->supportBackend(backend) || l->preferableTarget != target) && !fused)
+            {
+                hasFallbacks = true;
+                std::cout << "FALLBACK: Layer [" << l->type << "]:[" << l->name << "] is expected to has backend implementation" << endl;
+            }
+        }
+        ASSERT_FALSE(hasFallbacks) << "Implementation fallbacks are not expected in this test";
+    }
+};
+
+TEST_P(Test_ONNX_conformance, Layer_Test)
+{
+    applyTestTag(CV_TEST_TAG_DNN_SKIP_ONNX_CONFORMANCE);
```

Comment:
"Skip" tags should be applied with some condition only.

Reply:
I think, these tests should be disabled by default and would be triggered only if CI explicitly set `--test_tag_enable=dnn_skip_onnx_conformance`(in some `Custom` build). How can I achieve that?

Reply:
Test tags or OpenCV tests have different philosophy...

Current OpenCV "accuracy" tests (without tags) run always and prevent adding regressions into "existed codebase".
So,
0. "Enabled" test set of launched tests should not be empty.
1. "Enabled" test set of launched tests should be green.
2. Incoming patches should not make tests red.
3. Future patches may enable more test cases.

---

Next, "skip" tags disables tests and related code which is not supported (due to HW or SW dependency) or broken.
Skip tags should not be applied for all tests at once.

> `applyTestTag(CV_TEST_TAG_DNN_SKIP_ONNX_CONFORMANCE);`

So, we should not disable these tests by default.

---

Tags are designed for filtering of enabled test cases:
- `--gtest_filter` can do the same, but parameter value would be too long. Also its value must be updated somewhere (e.g., on CI which is not very convenient)
- tag-based filtering rules work in runtime, e.g. we can check HW requirements
- tag-based filtering rules work in runtime, e.g. we can check SW requirements - e.g. versions of used external libraries
- "skip" tests are smart variant of GTest `DISABLED_`

---

As we have all-in-one function we need to store "skip / unsupported cases" somewhere.
I would propose to have dedicated files for these rules:
- "parser" skip rules (common rules for all backends): file `test_onnx_conformance_onnx_layer_filter_parser.inl.hpp`
- several backend-specific skip rules: e.g. files  `test_onnx_conformance_onnx_layer_filter__{opencv,halide,ngraph,vulcan,cuda}.inl.hpp`

```
TEST_P(Test_ONNX_conformance, Layer_Test)
{
    std::string name = ...;
    Backend backend = ...;
    Backend target = ...;

    #include "test_onnx_conformance_onnx_layer_filter_parser.inl.hpp"
    if (backend == DNN_BACKEND_OPENCV)
    {
        #include "test_onnx_conformance_onnx_layer_filter__opencv.inl.hpp"
    }
#ifdef HAVE_HALIDE
    else if (backend == DNN_BACKEND_HALIDE)
    {
        #include "test_onnx_conformance_onnx_layer_filter__halide.inl.hpp"
    }
#endif
#ifdef HAVE_INF_ENGINE
    else if (backend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
    {
        #include "test_onnx_conformance_onnx_layer_filter__ngraph.inl.hpp"
    }
#endif
#ifdef HAVE_VULKAN
    else if (backend == DNN_BACKEND_VKCOM)
    {
        #include "test_onnx_conformance_onnx_filter__vulkan.inl.hpp"
    }
#endif
#ifdef HAVE_CUDA
    else if (backend == DNN_BACKEND_CUDA)
    {
        #include "test_onnx_conformance_onnx__cuda.inl.hpp"
    }
#endif
    else
    {
        std::ostringstream ss;
        ss << "No test filter available for backend ";
        PrintTo(backend, &ss);
        ss << ". Run test by default";
        std::cout << ss.str() << std::endl;
    }

    ... test code with loading model, inputs, forward, check results ...
}
```

We can use script which discovers tests (and generates `testConformanceConfig[]`) to generate these stubs for tests filtering:

```
else if (name == "test_abs")
{
    // execute test
}
else if (name == "test_acos")
{
    // execute test
}
...
```

So parser modifications would be simple (`test_onnx_conformance_onnx_filter_parser.inl.hpp`):
```
#define SKIP_TAGS \
    CV_TEST_TAG_DNN_SKIP_ONNX_PARSER, \
    CV_TEST_TAG_DNN_SKIP_ONNX_CONFORMANCE
#define SKIP applyTestTag(SKIP_TAGS)
#define SKIP_(...) applyTestTag(__VA_ARGS__, SKIP_TAGS)

else if (name == "test_abs")
{
    // execute test
}
else if (name == "test_acos")
{
    SKIP;  // optional comment with issue number
}
...
else
{
    ADD_FAILURE() << "Parser: unknown test='" << name << "'. Update filter configuration";
}

#undef SKIP_TAGS
#undef SKIP
#undef SKIP_
```

Backend filters would be similar but they would have more SKIP* macros:
- `SKIP` (all target devices): SKIP_(tag_target_skip)
- `SKIP_CPU`: `if (target == DNN_TARGET_CPU`) SKIP_(tag_target_skip)`
- `SKIP_NON_CPU`: `if (target != DNN_TARGET_CPU`) SKIP_(tag_target_skip)`
- `SKIP_OPENCL`
- `SKIP_OPENCL_FP16`
- CUDA-speficic: SKIP_CUDA

where variable `tag_target_skip` is initialized in filter's prolog with one of the following values:
- `CV_TEST_TAG_DNN_SKIP_CPU`
- `CV_TEST_TAG_DNN_SKIP_OPENCL`
- `CV_TEST_TAG_DNN_SKIP_OPENCL_FP16`
- ...

---

TBD later:
- how to update filter rules (how should we know that we need updates?)
- how to run all tests?
- how to generate conformance reports? (can we reuse existed regular test runs? - we can if our filters are up-to-date)
- checkBackend() improvement

---

I believe you could start with PARSER and OPENCV filters (CPU and/or OPENCL/OPENCL_FP16).
Other filters would be added later (or even generated!).
Note: CUDA filter may depend on CUDNN version.

Reply:
Updated PR.
Also updated PR's description.

Check opencv_extra's PR how to handle XML files (e.g., for reports).
Feel free to ask if you have any questions.

Reply:
[replaced](https://github.com/opencv/opencv/compare/8d3999b0d7e0fc04f651976b4dfc6c48b5439c76..8c0dd3126a737beca993e1005d340644971407cb) `else if() {...}` -> `if () { ... goto eof_label; }` in filters